internal/collector/collector.go

```diff
@@ -351,80 +434,116 @@ func generateCostManagementReports(log gologr.Logger, c *PrometheusCollector, di
+// generateCostVMMetricsReport generates the report for VM metrics.
+func generateCostVMMetricsReport(log gologr.Logger, c *PrometheusCollector, dirCfg *dirconfig.DirectoryConfig, yearMonth string) error {
+	log.Info("querying for cost vm metrics")
+	virtualMachineResults := mappedResults{}
+	if err := c.getQueryRangeResults(vmQueries, &virtualMachineResults, MaxRetries); err != nil {
 		return err
 	}
 
-	for vm, val := range vmResults {
+	virtualMachineRows := make(mappedCSVStruct)
+	for vm, val := range virtualMachineResults {
 		resourceID := getResourceID(val["provider_id"])
-		vmResults[vm]["resource_id"] = resourceID
-	}
+		virtualMachineResults[vm]["resource_id"] = resourceID
 
-	vmRows := make(mappedCSVStruct)
-	for vm, val := range vmResults {
 		usage := newVMRow(c.TimeSeries)
-		if err := getStruct(val, &usage, vmRows, vm); err != nil {
+		if err := getStruct(val, &usage, virtualMachineRows, vm); err != nil {
 			return err
 		}
 	}
+
 	emptyVmRow := newVMRow(c.TimeSeries)
-	vmReport := report{
+	virtualMachineReport := report{
 		file: &file{
 			name: vmFilePrefix + yearMonth + ".csv",
 			path: dirCfg.Reports.Path,
 		},
 		data: &data{
-			queryData: vmRows,
+			queryData: virtualMachineRows,
 			headers:   emptyVmRow.csvHeader(),
 			prefix:    emptyVmRow.dateTimes.string(),
 		},
 	}
-	log.WithName("writeResults").Info("writing virtual machine results to file", "filename", vmReport.file.getName())
-	if err := vmReport.writeReport(); err != nil {
-		return fmt.Errorf("failed to write virtual machine report: %v", err)
+
+	log.WithName("writeResults").Info("writing cost vm results to file", "filename", virtualMachineReport.file.getName())
+	if err := virtualMachineReport.writeReport(); err != nil {
+		return fmt.Errorf("failed to write cost vm report: %v", err)
 	}
 
-	//################################################################################################################
+	return nil
+}
 
-	log.Info("querying for namespaces")
-	namespaceResults := mappedResults{}
-	if err := c.getQueryRangeResults(namespaceQueries, &namespaceResults, MaxRetries); err != nil {
+// generateCostNvidiaGpuMetricsReport generates the report for NVIDIA GPU metrics.
+func generateCostNvidiaGpuMetricsReport(log gologr.Logger, c *PrometheusCollector, dirCfg *dirconfig.DirectoryConfig, yearMonth string) error {
+	log.Info("querying for cost nvidia gpu metrics")
+	nvidiaGpuResults := mappedResults{}
+	if err := c.getQueryRangeResults(costNvidiaGpuQueries, &nvidiaGpuResults, MaxRetries); err != nil {
 		return err
 	}
 
-	namespaceRows := make(mappedCSVStruct)
-	for namespace, val := range namespaceResults {
-		usage := newNamespaceRow(c.TimeSeries)
-		if err := getStruct(val, &usage, namespaceRows, namespace); err != nil {
+	nvidiaGpuRows := make(mappedCSVStruct)
+	gpuUtilizationResults := make(map[string]mappedValues)
+	gpuResourceResults := make(map[string]mappedValues)
+
+	// Separate the results
```

Comment:
My only real thought here is that this seems quite inefficient. We run the query, get all the results, split the results, and then re-consolidate the results.

Why not just run the queries separately and combine them afterwards?